astropy/utils/console.py

```diff
@@ -899,73 +899,80 @@ def __init__(self, msg, color='default', file=None, step=1,
         self._msg = msg
         self._color = color
         self._file = file
-        self._step = step
+        self._step = step + 1
+        self._cur_step = 0
         if chars is None:
             if conf.unicode_output:
                 chars = self._default_unicode_chars
             else:
                 chars = self._default_ascii_chars
         self._chars = chars
-
+        self._cur_char = 0
         self._silent = not isatty(file)
 
+        if self._silent:
+            self._iter = self._silent_iterator
+        else:
+            self._iter = self._iterator
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self, exc_type, exc_value, traceback):
+        file = self._file
+        write = file.write
+        flush = file.flush
+
+        if not self._silent:
+            write('\r')
+            color_print(self._msg, self._color, file=file, end='')
+        if exc_type is None:
+            color_print(' [Done]', 'green', file=file)
+        else:
+            color_print(' [Failed]', 'red', file=file)
+        flush()
+
+    def __iter__(self):
+        return self
+
+    def __next__(self):
+        self._iter()
+
     def _iterator(self):
+        """
+        Update the spin wheel.
+        """
         chars = self._chars
-        index = 0
         file = self._file
         write = file.write
-        flush = file.flush
         try_fallback = True
 
-        while True:
+        if self._cur_step % self._step == 0:
```

Comment:
I don't think changing this to not be a generator was necessary.  My thinking was in `__init__` you can do: `self._iter = self._iterate()`, and in `__next__` write:  `next(self._iter)`.

That said, you already went to the effort of changing it, so as long as it works I think it's a pretty harmless change.

Reply:
Uff. I thought I tried that approach, and at the time thought it didn't work. Though I just tried exactly what you wrote there and it works fine, so now I feel silly

Reply:
Don't worry about it; the way you rewrote it is fine too :)